api/src/main/java/org/openmrs/api/impl/UserServiceImpl.java

```diff
@@ -691,9 +713,91 @@ public String getSecretQuestion(User user) throws APIException {
 		if (user.getUserId() != null) {
 			LoginCredential loginCredential = dao.getLoginCredential(user);
 			return loginCredential.getSecretQuestion();
-		} else {
+		} 
+		else {
 			return null;
 		}
     }
 	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByUsernameOrEmail(java.lang.String)
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByUsernameOrEmail(String usernameOrEmail) {
+		if (StringUtils.isNotBlank(usernameOrEmail)) {
+			User user = dao.getUserByEmail(usernameOrEmail);
+			if (user == null) {
+				return getUserByUsername(usernameOrEmail);
+			}
+			return user;
+		}
+		throw new APIException("error.usernameOrEmail.notNullOrBlank", (Object[]) null);
+	}
+	
+	/**
+	 * @see org.openmrs.api.UserService#getUserByActivationKey(java.lang.String)
+	 * 
+	 */
+	@Override
+	@Transactional(readOnly = true)
+	public User getUserByActivationKey(String activationKey) {
+		LoginCredential loginCred = dao.getLoginCredentialByActivationKey(activationKey);
+		if (loginCred != null) {
+			String[] credTokens = loginCred.getActivationKey().split(":");
+			if (System.currentTimeMillis() <= Long.parseLong(credTokens[1])) {
+				return getUser(loginCred.getUserId());
+			}
+		}
+		return null;
+	}
+	
+	/**
+	 * @throws MessageException
+	 * @see org.openmrs.api.UserService#setUserActivationKey(org.openmrs.User)
+	 */
+	@Override
+	public User setUserActivationKey(User user) {
+		String token = RandomStringUtils.randomAlphanumeric(20);
+		long time = System.currentTimeMillis() + getValidTime();
+		String hashedKey = Security.encodeString(token);
+		String activationKey = hashedKey+":"+time;
+		LoginCredential credentials = dao.getLoginCredential(user);
+		credentials.setActivationKey(activationKey);	
+		dao.setUserActivationKey(credentials);	
+		
+		String link = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_HOST_URL) + token;
+		String msg = "Hi " + user.getUsername()
+		        + ", \n It looks like you like to change your password, click on the following link to do so.\n" + link
+		        + " please disregard this email if you didn't innitiate the password reset process  The openmrs Team";
+		Message mail = new Message();
+		mail.setId(20);
+		mail.setRecipients(user.getEmail());
+		mail.setSender(Context.getAdministrationService().getGlobalProperty("mail.from"));
+		mail.setContentType(Context.getAdministrationService().getGlobalProperty("mail.default_content_type"));
+		mail.setSubject("Password Reset Request");
+		mail.setContent(msg);
+		mail.setSentDate(new Date());
+		
+		try {
+			Context.getMessageService().sendMessage(mail);
+		}
+		catch (MessageException e) {
+			// Prevent propagation of exception to the webservice should in case message was not sent  
+		}
+		return user;
```

Comment:
So up to now, How the method will indicate that the email was sent without any faults to the caller? Do we want to indicate the caller about the email success or not?

Reply:
Users needs not know of the success or failure of the mails

Reply:
So if someone did not get the mail, then how can the user or System Administrator will get to know about that?
Is it good to have a debug log about this?